transaction/std/src/memo_builder/gift_code_memo_builder.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Defines the Gift Code Memo Builder for gift code memos specified in MCIP #32
+
+use super::{memo::UnusedMemo, MemoBuilder};
+use crate::{
+    memo::{GiftCodeCancellationMemo, GiftCodeFundingMemo, GiftCodeSenderMemo},
+    ChangeDestination,
+};
+use mc_account_keys::PublicAddress;
+use mc_crypto_keys::RistrettoPublic;
+use mc_transaction_core::{Amount, MemoContext, MemoError, MemoPayload, NewMemoError};
+use std::convert::TryFrom;
+
+/// The three possible memo types that can be written as specified in MCIP #32
+/// | Memo type bytes | Name                                              |
+/// | -----------     | -----------                                       |
+/// | 0x0002          | Gift Code Sender Memo                             |
+/// | 0x0201          | Gift Code Funding Memo                            |
+/// | 0x0202          | Gift Code Cancellation Memo                       |
+#[derive(Clone, Debug)]
+pub enum GiftCodeMemoType {
+    /// Gift Code Sender Memo 0x0002
+    Sender,
+    /// Gift Code Funding Memo 0x0201
+    Funding,
+    /// Gift Code Cancellation Memo 0x0202
+    Cancellation,
+}
+
+/// This memo builder writes gift code memos to the change address which can be
+/// used by apps that offer gift codes as feature to track the status of gift
+/// codes as well as certain contextual information about them. This builder
+/// writes 3 possible gift code memo types. Only one memo which can be specified
+/// at the creation of the memo builder. It is worth noting that all memos
```

Comment:
Ah yeah that's actually a great catch - I think as written, there' was a little bit of ambiguity around this, but I think that you're right that 4 distinct memos need to be written and one is destined for the change address. Thanks for catching that error.

Because there's also:
>Any TxOut's on the default subaddress are in-bound transactions, which may have 0x01.. authenticated sender memos which can be validated.
>Any TxOut's on the change subaddress contain recoverable transaction history memos that we can trust
> - 0x0200 Destination memo indicates an outbound transfer, and includes the recipient and the amount
> - 0x0201 Gift code funding memo indicates that we funded a gift code, and indicates which TxOut is the gift code
> - 0x0202 Gift code cancellation memo indicates that we cancelled a gift code that we previously funded
> - 0x0002 Gift code sender memo indicates that a gift code was claimed by the recipient and was sent by the sender
> - 0x0000 Unused indicates that we claimed a gift code but we aren't sure who it came from

Which leaves it somewhat amb

Reply:
Yeah but directly below that it states - stating that it must go to the change output.

> When claiming a gift code, the gift code output is spent in a self-payment to the change subaddress. This transaction has a single output, the change output, which has a memo of type 0x0002 Gift code sender memo.

Reply:
```mermaid
graph LR
    subgraph Gift
    B[Gift code output]
    C[Change output]
    end
    A[Funds] -->|TxOut| B
    A --> |Change| C
    C --> A
    B --> |TxOut| D(Claim)
```
(so cool that mermaid works in PR comments)
It may be my lack of understanding here.  It seems that there is an intermediate `TxOut` here, the "Gift code output" in the diagram.  My understanding is 
* "Change output" gets the funding memo.
* "Gift code output" gets the sender memo.
* "Claim" would get the cancellation memo, of the originator claims it.

I'm think `make_memo_for_change_output()` will only be called for the "Change output" and the self claim.  TBH not 100% on the self claim, it will be to the change subaddress, but does that force it through the entire change output logic...

Reply:
Mermaid JS in github is pretty awesome right? We should definitely be doing this more on MCIPs.

Here's roughly what the MCIP specifies
```mermaid
graph RL
    subgraph Sender_Account
    subgraph Sender_Gift_Code_Subaddress
    C[Gift_Code_TxOut]
    end
    subgraph Sender_Change_Address
    D[Zero_Value_Gift_Code_Funding_Memo_TxOut]
    E[Zero_Value_Gift_Cancellation_TxOut_with_Memo]
    end
    subgraph Sender_Default_Subaddress
    A[Available_TxIns]
    B[Cancellation_TxOut]
    end
    end
    subgraph Receiver_Account
    subgraph Receiver_Change_Address
    F[Redeemed_Gift Code_TxOut] 
    end
    subgraph Receiver_Gift_Code_Address
    G[Redeemed_Gift Code_TxOut] 
    end
    end

A -->|Funding - Funds| C
A -->|Funding - Memo| D
C -->|Cancellation - Funds| B
C -->|Cancellation - Memo| E
C -->|Redemption - Funds| G
C -->|Redemption - Memo| F
```

Reply:
I don't think the `Redemption - Funds` should be going to the gift code sub-address.  My understanding is that should only be used for _outbound_ in flight gifts.

I think the MCIP has some contradictions in it, as you mention it says:
> When cancelling a gift code, the gift code output is paid back to self. This transaction has a single output, the change output, which goes to the *change subaddress*

and
> When claiming a gift code, the gift code output is spent in a self-payment to the change subaddress. This transaction has a single output, the change output, which has a memo of type 0x0002 Gift code sender memo

However up in the guide level explanation it says:
> Until it is spent, the app can potentially cancel the gift code by paying the TxOut back to themself on subaddress 0.

and in the reference level explanation
> The app may cancel a gift code by paying it back to the primary subaddress.

Reply:
I wonder if it actually matters where the cancellation TxOut goes to (subaddress 0/primary or the change subaddress), since the way you tell it is a gift code cancellation is by looking at the memo. The only reason why I think having it go to the change subaddress is preferable, is because that subaddress is private so no-one could mess with your history.

Reply:
> I don't think the Redemption - Funds should be going to the gift code sub-address. My understanding is that should only be used for outbound in flight gifts.

Ya you're actually correct about this, it goes to the change address.

As far as the cancellation memo ambiguity, I think paying it back to the primary subaddress makes more sense given transaction builders can then use it in future txout selection.

Reply:
The transaction builder can use txouts from the change subaddress, it doesn't care which subaddress the outputs went to (if it couldn't then change would not be spendable...).

Reply:
I pull a modification to the MCIP to reflect this.